Update combined_traceback to work with 3.11 and 3.12 python.

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #170238

Summary:
This adds support for having multiple python frames correspond to one interpreter eval frame.

zdevito committed:
Update combined_traceback to work with 3.11 and 3.12 python.

Summary:
This adds support for having multiple python frames correspond to one interpreter eval frame.

[ghstack-poisoned]

albanD commented on test/test_utils.py:
```diff
@@ -987,6 +987,104 @@ def test_captured_traceback_format_all_cached(self):
+
+        # On Python 3.11+, the implementation should properly interleave frames
+        if sys.version_info >= (3, 11):
+            # All three of our functions should be visible in the trace
```
All frames are always visible since we append them at the bottom when not interleaved. We should check that they are actually properly interleaved

albanD reviewed (commented):
Sounds pretty fair.

zdevito committed:
Update on "Update combined_traceback to work with 3.11 and 3.12 python."


Summary:
This adds support for having multiple python frames correspond to one interpreter eval frame.

[ghstack-poisoned]

zdevito committed:
Update on "Update combined_traceback to work with 3.11 and 3.12 python."


Summary:
This adds support for having multiple python frames correspond to one interpreter eval frame.

[ghstack-poisoned]

zdevito committed:
Update on "Update combined_traceback to work with 3.11 and 3.12 python."


Summary:
This adds support for having multiple python frames correspond to one interpreter eval frame.

[ghstack-poisoned]

zdevito committed:
Update on "Update combined_traceback to work with 3.11 and 3.12 python."


Summary:
This adds support for having multiple python frames correspond to one interpreter eval frame.

[ghstack-poisoned]

sraikund16 reviewed (approved):
Overall LGTM. Thanks!

zdevito commented:
@pytorchbot merge

albanD reviewed (approved):
Sounds good!
Thanks for taking the time to do that!
